### PR TITLE
V07 snapshots

### PR DESCRIPTION
**Pull request by brennanjl**
This is a little bit hacky, but I believe it is the best we can do. This _should_ work for network migrations, however there could potentially be edge cases. I think I need to spend some more time thinking about how this can be done in the future, but as far as this release goes, it should work.

It ended up being less code to write that I thought, but it does feel extremely messy and hard to maintain long term.

### CODE REVIEW

File: internal/engine/execution/schema_v07.go [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+	"encoding/json"
+
+	"github.com/kwilteam/kwil-db/core/types"
+)
+
+// this file contains the schema from Kwil v0.7, which is used for proper migration to
+// the new schema in v0.8
+
+// convertV07Schema converts a v0.7 schema to a v0.8 schema
+func convertV07Schema(bts []byte) (*types.Schema, error) {
```
Comment (jchappelow) [inline internal/engine/execution/schema_v07.go:13]: This function and this file look fairly straightforward.  Some suggestions though: https://github.com/jchappelow/kwil-db/commit/15607b09c86bb3b897098d6f4df28cef7e0e8548
The only thing that looks like a bug is the "bool" vs. "BOOLEAN" change in the data type string.  Also case for all the types, but the code uses `strings.ToLower` a lot so case probably wouldn't matter.

File: internal/engine/execution/queries.go [hunk 2]
```diff
@@ -263,13 +263,22 @@ func createSchema(ctx context.Context, tx sql.TxMaker, schema *types.Schema, txi
...
+// getSchemas returns all schemas in the kwil_schemas table.
+// convertFunc converts bytes into a schema. If nil, it will simply unmarshal the bytes.
+func getSchemas(ctx context.Context, tx sql.Executor, convertFunc func([]byte) (*types.Schema, error)) ([]*types.Schema, error) {
 	res, err := tx.Execute(ctx, sqlListSchemaContent)
 	if err != nil {
 		return nil, err
 	}
 
+	if convertFunc == nil {
+		convertFunc = func(b []byte) (*types.Schema, error) {
```
Comment (jchappelow) [inline internal/engine/execution/queries.go:275]: Related concept, not requesting any changes, just food for thought: https://github.com/jchappelow/kwil-db/commit/423cdc6becfbe6690a5e054a9b4405f206345365
At least within the framework of JSON schema content, there are some simple ways to support reading prior schema versions if the schema doc and struct include a flag.  I suppose the only format we want to support migrating directly from text is KF, but we also support specifying a schema with JSON from kwil-cli so this is an exposed document structure with no real end-user accessible conversion.

Do we want to ditch the JSON option?  `  -t, --type string   file type of the database definition file (kf or json) (default "kf")`

Comment (brennanjl) [reply internal/engine/execution/queries.go:275]: I'm a bit confused, why would we need to ditch the json option in the CLI? It's just used to submit an already parsed schema. I don't see how it is related at all, other than that schemas parsed with v0.7 will not be readable by the CLI

**Comment by charithabandi** [comment]
Screenshot 2024-06-07 at 3 32 31 PM
I see these debugs. need to figure out why. Not sure if something is the issue with the voters/validators.

File: internal/engine/execution/queries.go (hunk 2 above)
Comment (jchappelow) [reply internal/engine/execution/queries.go:275]: It seems equally valid to provide your schema as an end user to the cli in kf or json.  Is it?

**Comment by charithabandi** [comment]
Ah! I had my node's private key wrong, looks good. Sorry for the alert.

File: internal/engine/execution/queries.go (hunk 2 above)
Comment (brennanjl) [reply internal/engine/execution/queries.go:275]: It is. They obtain the json by using the parser though. I don't think we have ever directed anyone to construct a schema via json, they simply get it from the IDE (which has an export parse to json) feature.

**Review by charithabandi — APPROVED** [review]

**Pull request merged**